src/lib.rs

```diff
@@ -56,6 +56,84 @@ where
     Err(())
 }
 
+#[cfg(target_env = "gnu")]
+mod glibc {
+    use std::{cmp, ffi::CStr, num, str};
+
+    #[derive(Debug, Eq, PartialEq, Copy, Clone)]
+    pub(crate) struct Version {
+        major: u16,
+        minor: u16,
+    }
+
+    impl Ord for Version {
+        fn cmp(&self, other: &Self) -> cmp::Ordering {
+            match self.major.cmp(&other.major) {
+                cmp::Ordering::Greater => return cmp::Ordering::Greater,
+                cmp::Ordering::Less => return cmp::Ordering::Less,
+                cmp::Ordering::Equal => (),
+            }
+            self.minor.cmp(&other.minor)
+        }
+    }
+
+    impl PartialOrd for Version {
+        fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
+            Some(self.cmp(other))
+        }
+    }
+
+    impl Version {
+        #[inline]
+        const fn new(major: u16, minor: u16) -> Self {
+            Self { major, minor }
+        }
+
+        pub(crate) fn from_libc() -> Self {
+            // SAFETY Call to libc via ffi.
+            let ptr = unsafe { libc::gnu_get_libc_version() };
```

Comment:
This appears to be checking `libc` on the current host that runs this "build script library", rather than testing if the headers/libraries intended for the _target_ have a certain version?

Reply:
This does check the host libc version. In case of static linking this works fine. In case of dynamic linking, from what I recall, the binary will be compiled with the host's glibc version and should not run on the target unless it has a equal or newer version of glibc.

> [...] rather than testing if the headers/libraries intended for the target have a certain version?

Isn't it the same? If you target gnu, your gonna need to compile with glibc.

Reply:
Here's few issues for reference on the equal or newer version requirement of glibc dynamic linking:
* https://github.com/rust-lang/rust/issues/57497
* https://github.com/rust-lang/rust-analyzer/issues/4426
* https://github.com/rust-lang/rust/issues/36826

Reply:
> In case of static linking this works fine

In case of static linking your host has the static library of the _target architecture_, however;

> then i think the approach is to compile and run a program that outputs the `libc::gnu_get_libc_version`.

This implies that your host can emulate the target architecture to run such a program.

---

The reason for the header check is because of the information about GLIBC in part comes from the sysroot.

Though afaik the sysroot/compiler is also available to set the GLIBC version and use older versions, in which case maybe the header emits the inline function as a fallback rather than the symbol? Or maybe older GLIBC header/library source is needed, I'm not too familiar with this.

Reply:
I'm also not too familiar with cross-compilation libc linking.

I guess your original solution would be a safe bet. We could try compiling a simple c program that calls `strlcpy` to emulate `CheckCXXSymbolExists`.

Reply:
I have a lot of experience with cross-compilation and building binaries that work in multiple glibc versions. I also worked in a project that runs on ARM and cross-compiles this crate. So this is my input:

* glibc is forward-compatible. If you compile for a lower target glibc version, then the same binary works in systems with higher glibc version.
* `libc::gnu_get_libc_version` gets the host glibc version, so cross-compilation will fail.
* The classic workaround that autotools and cmake are using, is to actually invoke gcc and the source code will be a small C code that just calls function in question. If the compilation succeeds then the function is supported, otherwise it doesn't.

Based on the above the best approach is what I describe in the 3rd point.

Reply:
Another options is to use `__GLIBC__` and `__GLIBC_MINOR__` and get their values with `cc::Build::expand`.

Reply:
@oblique Thanks for confirming the above with additional context. Indeed, it should also be possible to extract the glibc version of te _target sysroot_ via such environment variables.

The compilation test in C/C++ build-scripts is likely used in case a symbol is guarded by extra conditionals? Or more likely, to be interoperable across multiple libc standard library providers. Instead of hardcoding versions across platforms (for multiple symbols), this compile-test would for example run naturally on Android cross target instead of having to hardcode what we did in #20. Not saying that we should convert everything thus far to that though, but it's something to keep in mind.

Reply:
I think the best is to be used when we are not sure. For example in Android we are sure that exists, so we don't need to check via compilation test.

Reply:
> For example in Android we are sure that exists, so we don't need to check via compilation test.

I haven't gone back very far at all though, but it seems very unlikely for anyone to go out of their way and build against even older versions of the NDK (which may still even contain the symbol).